libyul/optimiser/UnusedStoreEliminator.h

```diff
@@ -0,0 +1,117 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Optimiser component that removes stores to memory and storage slots that are not used
+ * or overwritten later on.
+ */
+
+#pragma once
+
+#include <libyul/ASTForward.h>
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/optimiser/UnusedStoreBase.h>
+
+#include <map>
+#include <vector>
+
+namespace solidity::yul
+{
+struct Dialect;
+struct AssignedValue;
+
+/**
+ * Optimizer component that removes sstore statements if they
+ * are overwritten in all code paths or never read from.
+ *
+ * The m_store member of UnusedStoreBase is only used with the empty yul string
+ * as key in the first dimension.
+ *
+ * Best run in SSA form.
+ *
+ * Prerequisite: Disambiguator, ForLoopInitRewriter.
+ */
+class UnusedStoreEliminator: public UnusedStoreBase
+{
+public:
+	static constexpr char const* name{"UnusedStoreEliminator"};
+	static void run(OptimiserStepContext& _context, Block& _ast);
+
+	explicit UnusedStoreEliminator(
+		Dialect const& _dialect,
+		std::map<YulString, SideEffects> const& _functionSideEffects,
+		std::map<YulString, ControlFlowSideEffects> _controlFlowSideEffects,
+		std::map<YulString, AssignedValue> const& _ssaValues,
+		bool _ignoreMemory
+	):
+		UnusedStoreBase(_dialect),
+		m_ignoreMemory(_ignoreMemory),
+		m_functionSideEffects(_functionSideEffects),
+		m_controlFlowSideEffects(_controlFlowSideEffects),
+		m_ssaValues(_ssaValues)
+	{}
+
+	using UnusedStoreBase::operator();
+	void operator()(FunctionCall const& _functionCall) override;
+	void operator()(FunctionDefinition const&) override;
+	void operator()(Leave const&) override;
+
+	using UnusedStoreBase::visit;
+	void visit(Statement const& _statement) override;
+
+	enum class Location { Storage, Memory };
```

Comment:
```suggestion
	using Location = SemanticInformation::Location;
```
and get rid of ``translateLocation``? Or do we expect ``SemanticInformation::Location`` to be extended to calldata or returndata?

Reply:
Similarly, maybe even for ``Effect``... ``applyOperation`` can just ignore ``None`` (or assert ``!= None``)... but also fine to not change either.

Reply:
I wanted to avoid the dependency...

Reply:
On the header? Ok, fair enough... two types and two translation functions aren't that low of a price for avoiding that, but fine with me :-).